node/subsystem-util/src/database.rs

```diff
@@ -0,0 +1,335 @@
+			debug_assert!(
+				self.is_indexed_column(col),
+				"Invalid configuration of database, column {} is not ordered.",
+				col
+			);
+		}
+
+		fn filter_tx(&self, transaction: &DBTransaction) {
+			debug_assert!({
+				for op in &transaction.ops {
+					if let DBOp::DeletePrefix { col, .. } = op {
+						self.filter_iter(*col);
+					}
+				}
+				true
+			})
+		}
+	}
+
+	impl<D: KeyValueDB> Database for DbAdapter<D> {
+		fn is_indexed_column(&self, col: u32) -> bool {
+			self.allowed_iter.contains(&col)
+		}
+	}
+
+	impl<D: KeyValueDB> KeyValueDB for DbAdapter<D> {
+		fn transaction(&self) -> DBTransaction {
+			self.db.transaction()
+		}
+
+		fn get(&self, col: u32, key: &[u8]) -> Result<Option<DBValue>> {
+			self.db.get(col, key)
+		}
+
+		fn get_by_prefix(&self, col: u32, prefix: &[u8]) -> Option<Box<[u8]>> {
+			self.filter_iter(col);
+			self.db.get_by_prefix(col, prefix)
+		}
+
+		fn write(&self, transaction: DBTransaction) -> Result<()> {
+			self.filter_tx(&transaction);
+			self.db.write(transaction)
+		}
+
+		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+			self.filter_iter(col);
+			self.db.iter(col)
+		}
+
+		fn iter_with_prefix<'a>(
+			&'a self,
+			col: u32,
+			prefix: &'a [u8],
+		) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+			self.filter_iter(col);
+			self.db.iter_with_prefix(col, prefix)
+		}
+
+		fn restore(&self, _new_db: &str) -> Result<()> {
+			unimplemented!("restore is unsupported")
+		}
+
+		fn io_stats(&self, kind: IoStatsKind) -> IoStats {
+			self.db.io_stats(kind)
+		}
+
+		fn has_key(&self, col: u32, key: &[u8]) -> Result<bool> {
+			self.db.has_key(col, key)
+		}
+
+		fn has_prefix(&self, col: u32, prefix: &[u8]) -> bool {
+			self.filter_iter(col);
+			self.db.has_prefix(col, prefix)
+		}
+	}
+
+	impl<D: KeyValueDB> MallocSizeOf for DbAdapter<D> {
+		fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
+			// ignore filter set
+			self.db.size_of(ops)
+		}
+	}
+}
+
+/// Utilities for using parity-db database.
+pub mod paritydb_impl {
+	use super::{DBTransaction, DBValue, Database, KeyValueDB};
+	use kvdb::{DBOp, IoStats, IoStatsKind};
+	use parity_db::Db;
+	use parking_lot::Mutex;
+	use std::{collections::BTreeSet, io::Result, sync::Arc};
+
+	fn handle_err<T>(result: parity_db::Result<T>) -> T {
+		match result {
+			Ok(r) => r,
+			Err(e) => {
+				panic!("Critical database error: {:?}", e);
+			},
+		}
+	}
+
+	fn map_err<T>(result: parity_db::Result<T>) -> Result<T> {
+		result.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", e)))
+	}
+
+	/// Implementation of of `Database` for parity-db adapter.
+	pub struct DbAdapter {
+		db: Db,
+		allowed_iter: BTreeSet<u32>,
+		write_lock: Arc<Mutex<()>>,
+	}
+
+	impl parity_util_mem::MallocSizeOf for DbAdapter {
+		fn size_of(&self, _ops: &mut parity_util_mem::MallocSizeOfOps) -> usize {
+			unimplemented!("size_of is not supported for parity_db")
```

Comment:
hmm, this might bite us in the future

Reply:
I think this comes from the requirement of KeyValueDB and is not actually used. So if we will go with a separate trait approach: https://github.com/paritytech/polkadot/pull/4838#discussion_r809084800, we won't need it.